hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerReplicaCount.java

```diff
@@ -22,238 +22,36 @@
-   * ensure the container is perfectly replicate. Inflight operations are not
-   * considered here, but the logic to determine the missing or excess counts
-   * for maintenance is present.
-   *
-   * Decisions around over-replication are made only on healthy replicas,
-   * ignoring any in maintenance. For example, if we have:
-   *
-   *     H, H, H, M, M
-   *
-   * This will not be consider over replicated until one of the Maintenance
-   * replicas moves to Healthy.
-   *
-   * If the container is perfectly replicated, zero will be return.
-   *
-   * If it is under replicated a positive value will be returned, indicating
-   * how many replicas must be added.
-   *
-   * If it is over replicated a negative value will be returned, indicating now
-   * many replicas to remove.
-   *
-   * @return Zero if the container is perfectly replicated, a positive value
-   *         for under replicated and a negative value for over replicated.
-   */
-  private int missingReplicas() {
-    int delta = repFactor - healthyCount;
-
-    if (delta < 0) {
-      // Over replicated, so may need to remove a container.
-      return delta;
-    } else if (delta > 0) {
-      // May be under-replicated, depending on maintenance.
-      delta = Math.max(0, delta - maintenanceCount);
-      int neededHealthy =
-          Math.max(0, minHealthyForMaintenance - healthyCount);
-      delta = Math.max(neededHealthy, delta);
-      return delta;
-    } else { // delta == 0
-      // We have exactly the number of healthy replicas needed.
-      return delta;
-    }
-  }
-
-  /**
-   * Return true if the container is sufficiently replicated. Decommissioning
-   * and Decommissioned containers are ignored in this check, assuming they will
-   * eventually be removed from the cluster.
-   * This check ignores inflight additions, as those replicas have not yet been
-   * created and the create could fail for some reason.
-   * The check does consider inflight deletes as there may be 3 healthy replicas
-   * now, but once the delete completes it will reduce to 2.
-   * We also assume a replica in Maintenance state cannot be removed, so the
-   * pending delete would affect only the healthy replica count.
-   *
-   * @return True if the container is sufficiently replicated and False
-   *         otherwise.
-   */
-  public boolean isSufficientlyReplicated() {
-    return missingReplicas() + inFlightDel <= 0;
-  }
-
-  /**
-   * Return true is the container is over replicated. Decommission and
-   * maintenance containers are ignored for this check.
-   * The check ignores inflight additions, as they may fail, but it does
-   * consider inflight deletes, as they would reduce the over replication when
-   * they complete.
-   *
-   * @return True if the container is over replicated, false otherwise.
-   */
-  public boolean isOverReplicated() {
-    return missingReplicas() + inFlightDel < 0;
-  }
+  int additionalReplicaNeeded();
```

Comment:
I wonder if we should omit this from the interface. For EC, its a tricky calculation and probably not that useful. It could say 2 additional replicas needed, but its not too helpful, as we don't know what indexes, or if its a reconstruction or an easy copy from a decommissioning node. It feels like this doesn't apply well to the common methods, and I don't think its used inside the decommission code.

Reply:
Yep, turns out it's not needed.